### PR TITLE
ENH PHP 8.1 compatibility

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/cow/issues/232

**Changes:**
_note: the checkboxes are so that the reviewer can more easily say what they've checked since changes aren't separated as separate commits_

- [ ] Require PHP 8.1
- [x] Update dependencies
- [ ] Adding TEST_MODE which prevents pushing to github and transifex
- [x] No longer use -release branches
- [ ] Create minor branches if required for beta and rc tags (though still not for alpha)
- [ ] Add in graphql hack
- [x] Use HEAD for "to-version" in audit changelog
- [x] Show post-rc commits in audit changelog

Requires https://github.com/silverstripe/gha-generate-matrix/pull/42 to be merged and released for unit tests to pass

### CODE REVIEW

**Comment by GuySartorelli** [comment]
CI still failing after releasing the gha-generate-matrix change and rerunning all jobs

**Comment by GuySartorelli** [comment]
prf-low and phplinting jobs failed

**Comment by GuySartorelli** [comment]
Next time you do a bunch of unrelated changes in a single PR like this _please please please_ add them as separate commits! It would make reviewing this _way_ less of a headache if I could review each piece of functionality at a time without having to figure out what's what.

**Comment by GuySartorelli** [comment]
The stabalisation changes aren't mentioned in the description??

**Comment by emteknetnz** [comment]
The stabilisation changes are covered by "No longer use -release branches" - I'm assuming the stabilisation was working correctly beforehand.

File: src/Steps/Release/MergeUpRelease.php [hunk 1]
```diff
@@ -78,16 +80,14 @@ protected function mergeRecursive(
...
-        $releaseBranch = $library->getBranch();
-        $minorBranch = str_replace(LibraryRelease::RELEASE_BRANCH_SUFFIX, '', $releaseBranch);
+        $minorBranch = $library->getBranch();
         $majorBranch = $releasePlanNode->getVersion()->getMajor();
 
-        // Merge up to minor branch, then to major branch
-        $this->mergeUp($output, $releasePlanNode, $releaseBranch, $releaseBranch, $minorBranch, $pushToRemote, false);
-        $this->mergeUp($output, $releasePlanNode, $releaseBranch, $minorBranch, $majorBranch, $pushToRemote, true);
+        // Merge up to major branch
+        $this->mergeUp($output, $releasePlanNode, $minorBranch, $majorBranch, $pushToRemote, true);
```
Comment (GuySartorelli) [inline src/Steps/Release/MergeUpRelease.php:87]: Is the `$isMajorBranch` parameter for `mergeUp()` needed anymore? I don't think it is?

File: src/Model/Release/LibraryRelease.php [hunk 2]
```diff
@@ -349,9 +347,9 @@ public function setVersion($version)
     /**
      * True if the current branch ends in the release branch suffix
      */
-    public function isOnReleaseBranch(): bool
+    public function isOnMinorBranch(): bool
```
Comment (GuySartorelli) [inline src/Model/Release/LibraryRelease.php:350]: We should probably implement this as `isOnMinorReleaseBranch()` or similarly named instead (i.e. it's on _the correct_ minor branch) and use the plan to determine the correct branch.

Though if you think that's more effort than it's worth I will defer to your judgment.

File: src/Model/Release/ComposerConstraint.php [hunk 3]
```diff
@@ -16,6 +16,13 @@
  */
 class ComposerConstraint
 {
+    /**
+     * Used to hack in dual support for graphql on CMS 4
+     * Update this whenever a new version of graphql 3 is released
+     */
+    private const GRAPHQL_3_VERSION_DEV = '3.8.x-dev';
+    private const GRAPHQL_3_VERSION_STABLE = '~3.8.1@stable';
```
Comment (GuySartorelli) [inline src/Model/Release/ComposerConstraint.php:24]: I feel like it would be too easy to forget to update this if we release a new patch.
How important is that `.1` there as opposed to `.0`? If we're going to specify the exact patch version I think we should get that from the github api instead of hardcoding it.

File: src/Model/Changelog/Changelog.php [hunk 4]
```diff
@@ -93,8 +129,7 @@ protected function getLibraryLog(OutputInterface $output, ChangelogLibrary $chan
      * @param ?callable $filter function for array_filter on the list of changes
      * @return ChangelogItem[]
      */
-    protected function getGroupedChanges(OutputInterface $output, ?callable $filter = null)
-    {
+    protected function getGroupedChanges(OutputInterface $output, ?callable $filter = null) {
```
Comment (GuySartorelli) [inline src/Model/Changelog/Changelog.php:132]: This change will fail linting

File: src/Model/Changelog/Changelog.php [hunk 5]
```diff
@@ -47,15 +47,24 @@ public function __construct(ChangelogLibrary $rootLibrary)
      * @param ChangelogLibrary $changelogLibrary
      * @return array
      */
-    protected function getLibraryLog(OutputInterface $output, ChangelogLibrary $changelogLibrary)
-    {
+    protected function getLibraryLog(
+        OutputInterface $output,
+        ChangelogLibrary $changelogLibrary,
+        ?bool $changelogAuditMode = null
```
Comment (GuySartorelli) [inline src/Model/Changelog/Changelog.php:53]: Why make this a tri-state instead of just default to `false`?

File: src/Commands/Command.php [hunk 6]
```diff
@@ -61,7 +61,10 @@ protected function execute(InputInterface $input, OutputInterface $output)
         // Configure extra output formats
         $this->output->getFormatter()->setStyle('bold', new OutputFormatterStyle('blue'));
 
-        return $this->fire();
+        $ret = $this->fire();
+        // returning a default value of 1 here to satisfy
+        // symfony/console/Command/Command.php::run()
+        return is_int($ret) ? $ret : 1;
```
Comment (GuySartorelli) [inline src/Commands/Command.php:67]: We should return one of the constants defined on `Command` - and for that matter we should probably just get `fire()` to return the appropriate const based on whether it was successful or not.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Commands/Command.php (hunk 6 above)
Comment (emteknetnz) [reply src/Commands/Command.php:67]: Have changed the default `1` to Console\Command\Command::SUCCESS

I don't think updating the 17 fire() methods is worth it, if we did this we'd also want to return `Console\Comman\Command::FAILURE` everytime there's some sort of error, so we'd need to update lots of code.

I really don't think slightly nicer output colours (I'm presuming that's what the outcome) is worth the effort.

Comment (GuySartorelli) [reply src/Commands/Command.php:67]: Yeah, fair enough. I think cow is a mangled enough mess that the benefits aren't really beneficial without a considerable refactor.

File: src/Model/Changelog/Changelog.php (hunk 5 above)
Comment (emteknetnz) [reply src/Model/Changelog/Changelog.php:53]: Have changed to regular `bool`

File: src/Model/Changelog/Changelog.php (hunk 4 above)
Comment (emteknetnz) [reply src/Model/Changelog/Changelog.php:132]: Updated

File: src/Model/Release/ComposerConstraint.php (hunk 3 above)
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php:24]: In practice it makes no difference.

Using the github api from ComposerConstraint would be nicer, but it requires a lot of code change to do this.
There's actually 2x github api clients
- GithubApi (old, seems unused)
- PublishRelease::getGithubClient() (the one that is actually used e.g. creating github releases)

Unfortunately PublishRelease::getGithubClient() cannot easily be separated (I tried) due to it needing to instantiate a StepCommandRunner, which itself needs a Step.  It's basically just needs a solid refactor in order to use it.

The chance of graphql 3.9 being released at this stage seems fairly remote.  And if it is, it'll be in the context of a security release, and I'd expect we'd be pretty conscious of checking the what's released in the CMS 4 version of recipe-cms at that stage.

Comment (GuySartorelli) [reply src/Model/Release/ComposerConstraint.php:24]: > In practice it makes no little difference if it's .1 or .0. It would only affect a --prefer-lowest CI build.

Let's set it to `.0` then to make the intention clearer (i.e. we don't care that patch version is, what we care about is that it's stable)

File: src/Model/Release/ComposerConstraint.php [hunk 7]
```diff
@@ -128,6 +135,16 @@ public function __construct($constraint, $parentVersion = null, $name = null)
             return;
         }
 
+        // HACK for graphql where there's dual support for 3.x-dev || 4.x-dev
+        // This may come also come in as something like 3.8.x-dev || 4.1.x-dev
+        // This temporarily converts it to something that's "semver compatible"
+        // This is only for CMS 4. CMS 5 will ignore this
+        if ($name == 'silverstripe/graphql') {
+            if (preg_match('#3[\.0-9]*\.x-dev \|\| (4[\.0-9]*\.x-dev)#', $constraint, $matches)) {
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php:143]: Only when it's .x-dev format. I've run this through an entire simlulated (DEV_MODE=1) release for 4.13.0, the tagged versions were what was expected for at each point.

File: src/Model/Release/LibraryRelease.php (hunk 2 above)
Comment (emteknetnz) [reply src/Model/Release/LibraryRelease.php:350]: I don't think it's worth it at this stage, I've run simulated (DEV_MODE=1) releases for 4.13.0 and 5.0.0-beta1 and thing seemed to work as they should.

File: src/Steps/Release/MergeUpRelease.php (hunk 1 above)
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:87]: Yes, it's passed to ConstraintDestabiliser::destabiliseConstraints() which does use it

File: src/Model/Release/LibraryRelease.php (hunk 2 above)
Comment (GuySartorelli) [reply src/Model/Release/LibraryRelease.php:350]: This suggestion is more to protect against _human_ error (e.g. I manually checkout `4.12` after resolving merge conflicts, but I should have typed `4.13`)

File: src/Steps/Release/MergeUpRelease.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:87]: Right, but that's just always going to be true for mergup, right? So we can remove the parameter from mergeup and just pass in true from there.

File: src/Steps/Release/RewriteReleaseBranches.php [hunk 8]
```diff
@@ -253,8 +257,8 @@ protected function getTargetBranch($targetVersion, $branching, $currentBranch)
             case Branch::MAJOR:
                 return $majorBranch;
             case Branch::AUTO:
-                // Auto disables branching for unstable tags
-                if (!$targetVersion->isStable()) {
+                // Auto disables branching for alpha tags
+                if (!$targetVersion->getStability() === 'alpha') {
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: Because new branches weren't being created for beta releases when the default "auto" branches was used.  Saves having to change the branching strategy to "minor" in the plan

Since we always start with a beta (we didn't used to) this change makes sense to me

File: src/Steps/Step.php [hunk 9]
```diff
@@ -109,15 +102,14 @@ public function runCommand(OutputInterface $output, $command, $error = null, $ex
 
         // Prepare unbound command
         if (is_array($command)) {
-            $process = ProcessBuilder::create($command)->getProcess();
+            $process = new Process($command);
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Steps/Step.php:105]: `Symfony\Component\Process\ProcessBuilder` was removed. Replacement was simply to use `new Process($command);`

File: src/Steps/Step.php [hunk 10]
```diff
@@ -109,15 +102,14 @@ public function runCommand(OutputInterface $output, $command, $error = null, $ex
...
         } elseif ($command instanceof Process) {
             $process = $command;
         } else {
-            $process = new Process($command);
+            $process = new Process([$command]);
         }
 
         // Set all default env vars
-        $process->inheritEnvironmentVariables();
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Steps/Step.php:120]: `[Symfony 5] The "Symfony\Component\Process\Process::inheritEnvironmentVariables()" method is deprecated since Symfony 4.4, env variables are always inherited`

Removed and no need for a replacement

File: src/Steps/Release/RewriteReleaseBranches.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: Ahhh okay that's what the branching strategy does. I assume this will come with a relevant change in the process doc as well.
Sounds good.

File: src/Steps/Release/PublishRelease.php [hunk 11]
```diff
@@ -92,24 +95,18 @@ protected function releaseLibrary(OutputInterface $output, LibraryRelease $relea
         // Step 1: Rewrite composer.json to all tagged versions only
         $this->stabiliseRequirements($output, $releasePlanNode);
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Steps/Release/PublishRelease.php:96]: Updated

File: src/Steps/Release/PublishRelease.php [hunk 12]
```diff
@@ -144,6 +141,9 @@ protected function stabiliseRequirements(OutputInterface $output, LibraryRelease
             );
         }
 
+        // HACK - update graphql version to dual support for CMS 4
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Steps/Release/PublishRelease.php:144]: Updated

File: src/Steps/Release/RewriteReleaseBranches.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: > Ahhh okay that's what the branching strategy does

I honestly don't know what it does and whatever it does it's probably busted.  I don't even know what I'd document.

File: src/Model/Release/LibraryRelease.php (hunk 2 above)
Comment (emteknetnz) [reply src/Model/Release/LibraryRelease.php:350]: Updated

File: src/Steps/Release/RewriteReleaseBranches.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: Oh okay. I thought you'd figured it out.
No worries then, so long as it's clear from the confluence doc what to do, I'm not fussed.

File: src/Model/Release/ComposerConstraint.php (hunk 3 above)
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php:24]: `.1` is better, it could make a difference on `--prefer-lowest` builds for CMS 4

When I say "in practice" - no-one other than our CI use --prefer-lowest.  It's still nice to have for our own purposes.

Comment (GuySartorelli) [reply src/Model/Release/ComposerConstraint.php]: Then I revert back to my original position. If we care about the patch version we should just the API to get the correct patch version, since we're not likely to remember to update that as we do patch releases.

File: src/Steps/Release/MergeUpRelease.php (hunk 1 above)
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:87]: mergeUp() is currently agnostic if it's merging minor->minor or minor->major.

I know that our current workflow is always minor->major, but they may change in the future so may as well just leave it agnostic.  The recommended change would force it to be minor->major (or major->major) ... however all the language and variable names within mergeUp() are written for it being agnostic

tl;dr just leave it as is

File: src/Model/Release/ComposerConstraint.php (hunk 3 above)
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php]: As mentioned above I think there's too much refactoring required at this stage to get this working.  We can open a new card if you think it's worthwhile to look at later

Comment (GuySartorelli) [reply src/Model/Release/ComposerConstraint.php]: This hasn't changed since last time it was assigned to me.... I'll just let it stay as is, it's not a big deal.

File: src/Model/Release/LibraryRelease.php [hunk 13]
```diff
@@ -347,11 +345,13 @@ public function setVersion($version)
...
     /**
-     * True if the current branch ends in the release branch suffix
+     * Check if the current minor branch matches what it expected to be given the new tag in the plan
      */
-    public function isOnReleaseBranch(): bool
+    public function isOnCorrectMinorReleaseBranch(): bool
     {
-        $branch = $this->getLibrary()->getBranch();
-        return preg_match('/' . preg_quote(self::RELEASE_BRANCH_SUFFIX) . '$/', $branch);
+        if (!preg_match('#^([0-9]{1,2}\.[0-9]{1,2})\.[0-9]{1,2}$#', $this->version, $matches)) {
```
Comment (GuySartorelli) [inline src/Model/Release/LibraryRelease.php:352]: `$this->version` is what's in the plan, right? Won't this then always return false if we're doing a beta or rc (where version has `-beta1` at the end for example).
If so, is that intentional? I'm trying to understand how this condition is helping us.

Comment (emteknetnz) [reply src/Model/Release/LibraryRelease.php]: Ah yep, have updated to remove the trailing '$' so it'll match beta1 etc

Main thing is to extra the major.minor so we get the expected branch name

Comment (GuySartorelli) [reply src/Model/Release/LibraryRelease.php]: Sweet. I don't think the condition is needed but it also doesn't hurt anything so I'm okay with it staying for this internal thing.

**Review by GuySartorelli — APPROVED** [review]
